### PR TITLE
add a plan cache Introduction document

### LINKED ISSUE #90: Write down the first version of Plan Cache section

### PR DESCRIPTION
**Pull request by qw4990**
### What issue does this PR solve?

- close #90

### What is changed:
Add an introduction to the plan cache. (WIP)

### CODE REVIEW

File: src/understand-tidb/plan-cache.md [hunk 1]
```diff
@@ -1 +1,59 @@
 # Plan Cache
+
+TiDB supports execution plan caching for `Prepare` and `Execute` queries. By using cached plan, TiDB can skip the optimization phase and then gain some performance benefits.
+
+The flow of handling `Prepare` and `Execution` is shown below:
+
+![plan-cache-overview](../img/plan-cache-overview.png)
```
Comment (tisonkun) [inline src/understand-tidb/plan-cache.md:7]: Could you remove the grid of the picture?

File: src/understand-tidb/plan-cache.md [hunk 2]
```diff
@@ -1 +1,59 @@
...
+
+There still are some limitations about TiDB plan cache, and please see [the TiDB Plan Cache Doc](https://docs.pingcap.com/tidb/stable/sql-prepare-plan-cache) for more details.
+
+# The flow of handling Prepare statement 
+
+The flow of handling `Prepare` statement is not complex, and the main work are below:
+1. parse all parameter placeholders, 
+2. parse the query into an AST,
+3. store them into current session's `PrepareStmtCache`.
```
Comment (tisonkun) [inline src/understand-tidb/plan-cache.md:16]: I don't find a symbol named `PrepareStmtCache`.

Comment (tisonkun) [inline src/understand-tidb/plan-cache.md:16]:
```suggestion
1. Parse all parameter placeholders.
2. Parse the query into an AST.
3. Store them into current session's `PrepareStmtCache`.
```

File: src/understand-tidb/plan-cache.md (hunk 1 above)
Comment (tisonkun) [reply src/understand-tidb/plan-cache.md:7]: Also there is no symbol `PrepStmt`

File: src/understand-tidb/plan-cache.md [hunk 3]
```diff
@@ -1 +1,59 @@
...
+
+The entry function is `clientConn.handleStmtPrepare`, and most work is done in `PrepareExec.Next`.
```
Comment (tisonkun) [inline src/understand-tidb/plan-cache.md:18]: For the cache part, I found other logic in

* common_plans.go: func (e *Execute) getPhysicalPlan(ctx context.Context, sctx sessionctx.Context, is infoschema.InfoSchema, preparedStmt *CachedPrepareStmt) error

where `sctx.PreparedPlanCache().Put(cacheKey, cacheVals)` etc. called.

File: src/understand-tidb/plan-cache.md [hunk 4]
```diff
@@ -1 +1,59 @@
...
+1. parse all execution parameters and get their values,
+2. store all parameter values into current session's param-value cache,
+3. check whether there is a cached plan for the current statement, and
+   1. if such a cached plan exists, then 
+      1. rebuild the cached plan, and
+      2. run the rebuilt cached plan;
+   2. else 
+      1. optimize the cached AST with current parameters and get a new plan,
+      2. store the new plan into current session's plan cache,
+      3. run the new plan.
```
Comment (tisonkun) [inline src/understand-tidb/plan-cache.md:34]: I think you can write a pseudo code snippet for this.

File: src/understand-tidb/plan-cache.md [hunk 5]
```diff
@@ -1 +1,59 @@
...
+
+The entry function of handling `Execute` statement is `clientConn.handleStmtExecution`, and most work is done in `Execute.getPhysicalPlan`.
+
+# The Paramater Maker
+
+This section explain how TiDB handle parameters and propagate them into the execution plan when processing.
```
Comment (tisonkun) [inline src/understand-tidb/plan-cache.md:40]:
```suggestion
This section explain how TiDB handles parameters and propagate them into the execution plan when processing.
```

**Review by tisonkun — COMMENTED** [review]
Thanks for creating this PR. Comments inline.

**Comment by qw4990** [comment]
Duplicated with https://github.com/pingcap/tidb-dev-guide/pull/199

**Pull request closed without merging**